xarray/core/combine.py

```diff
@@ -605,21 +560,141 @@ def auto_combine(datasets, concat_dim=_CONCAT_DIM_DEFAULT,
-                         data_vars=data_vars, coords=coords,
-                         infer_order_from_coords=infer_order_from_coords,
-                         ids=False)
+        message = """In xarray version 0.14 `auto_combine` will be deprecated.
+                  The datasets supplied do not have global dimension
+                  coordinates. In future, to continue concatenating without
+                  supplying dimension coordinates, please use the new
+                  `combine_manual` function (or the `combine='manual'` option
+                  to open_mfdataset)."""
+        warnings.warn(message, FutureWarning)
+
+    if _requires_concat_and_merge(datasets):
+        manual_dims = [concat_dim].append(None)
+        message = """In xarray version 0.14 `auto_combine` will be deprecated.
+                  The datasets supplied require both concatenation and merging.
+                  From xarray version 0.14 this will operation will require
+                  either using the new `manual_combine` function (or the
+                  `combine='manual'` option to open_mfdataset), with
+                  a nested list structure such that you can combine along the
+                  dimensions {}. Alternatively if your datasets have global
+                  dimension coordinates then you can use the new `combine_auto`
+                  function.""".format(manual_dims)
+        warnings.warn(message, FutureWarning)
+
+    return _old_auto_combine(datasets, concat_dim=concat_dim,
+                             compat=compat, data_vars=data_vars,
+                             coords=coords)
+
+
+def _dimension_coords_exist(datasets):
+    """
+    Check if the datasets have consistent global dimension coordinates
+    which would in future be used by `auto_combine` for concatenation ordering.
+    """
+
+    # Group by data vars
+    sorted_datasets = sorted(datasets, key=vars_as_keys)
+    grouped_by_vars = itertools.groupby(sorted_datasets, key=vars_as_keys)
+
+    # Perform the multidimensional combine on each group of data variables
+    # before merging back together
+    concatenated_grouped_by_data_vars = []
+    try:
+        for vars, datasets_with_same_vars in grouped_by_vars:
+            _infer_concat_order_from_coords(list(datasets_with_same_vars))
+        return True
+    except ValueError as err:
+        no_dimension_coords_errs = ["Every dimension needs a coordinate",
+                                    "neither monotonically increasing nor",
+                                    "Cannot handle size zero",
+                                    "Could not find any dimension coordinates"]
+        if any(message in str(err) for message in no_dimension_coords_errs):
+            # The ValueError just means that the datasets don't have
+            # global dimension coordinates
+            return False
+        else:
+            # There is a different problem
+            raise err
```

Comment:
Given that this is an internal check only used for determining the right deprecation warning, it might be better to swallow this error instead.

Reply:
Sorry I'm not sure exactly what you mean... If we don't raise any remaining errors here then the code will just continue until it errors in some less explicit way?

Reply:
This helper function is only used inside `auto_combine` as part of figuring out the proper deprecation warning. So the conservative thing to do would be to issue some sort of "default" deprecation warning if something unexpected happens, rather than to raise a new error.

In particular, I'm a little concerned that your checks above in `no_dimension_coords_errs` might not always do the right thing -- they are matching strings in error messages, which can be fragile. If `ValueError` is raised for another reason, the user could get a surprising traceback.

Reply:
> issue some sort of "default" deprecation warning if something unexpected happens

I could raise a default deprecation warning, then if no errors are raised while trying to get through `_dimension_coords_exist()` raise another warning with more info? Then the user will always get a deprecation warning regardless of what error is thrown?

> matching strings in error messages, which can be fragile

Okay, but what other way is there of doing this? We have to distinguish between these errors if we want to be able to check what type of input the user gave and still carry on going.

> If ValueError is raised for another reason, the user could get a surprising traceback.

If `ValueError` is raised for another reason then the user will see that because it will be caught and raised, no?

Do you mean that I should raise the particular error caught by `_infer_concat_order_from_coords` as a second warning?

Reply:
> I could raise a default deprecation warning, then if no errors are raised while trying to get through `_dimension_coords_exist()` raise another warning with more info? Then the user will always get a deprecation warning regardless of what error is thrown?

This sounds like a good solution to me.

> Okay, but what other way is there of doing this? We have to distinguish between these errors if we want to be able to check what type of input the user gave and still carry on going.

The alternative is either (1) making dedicated error classes or (2) refactoring into lower level functions with more specific functionality. That said, it's probably not worth bothering with either of these here.


> > issue some sort of "default" deprecation warning if something unexpected happens
> 
> I could raise a default deprecation warning, then if no errors are raised while trying to get through `_dimension_coords_exist()` raise another warning with more info? Then the user will always get a deprecation warning regardless of what error is thrown?
> 
> > matching strings in error messages, which can be fragile
> 
> Okay, but what other way is there of doing this? We have to distinguish between these errors if we want to be able to check what type of input the user gave and still carry on going.
> 
> > If ValueError is raised for another reason, the user could get a surprising traceback.
> 
> If `ValueError` is raised for another reason then the user will see that because it will be caught and raised, no?
> 
> Do you mean that I should raise the particular error caught by `_infer_concat_order_from_coords` as a second warning?

I think it's better just to silence the error.